sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/Error.scala

```diff
@@ -71,6 +75,27 @@ object Error {
           s"the missing dependencies are ${missingDependencies.mkString("{'", "', '", "'}")}."
     }
 
+    final case class DarSelfConsistency(
+        mainPackageId: Ref.PackageId,
+        transitiveDependencies: Set[Ref.PackageId],
+        missingDependencies: Set[Ref.PackageId],
+        extraDependencies: Set[Ref.PackageId],
+    ) extends Error
+        with OptionalLogReporting {
+      def message: String =
+        s"For package $mainPackageId, the set of package dependencies ${transitiveDependencies
+            .mkString("{'", "', '", "'}")} is not self consistent, " +
+          (if (missingDependencies.nonEmpty)
+             s"the missing dependencies are ${missingDependencies.mkString("{'", "', '", "'}")} "
+           else "") +
+          (if (missingDependencies.nonEmpty && extraDependencies.nonEmpty) "and " else "") +
+          (if (extraDependencies.nonEmpty)
+             s"the extra dependencies are ${extraDependencies.mkString("{'", "', '", "'}")}"
+           else "")
+
+      override def logReportingEnabled: Boolean =
```

Comment:
When the participant is configured to log (rather than report a Grpc error), this function determine when logging should actually occur. That way the engine controls when an error is logged and when it is returned as a Grpc error.